go/store/prolly/tree/immutable_tree.go

```diff
@@ -161,27 +166,72 @@ func NewByteArray(addr hash.Hash, ns NodeStore) *ByteArray {
 }
 
 func (b *ByteArray) ToBytes(ctx context.Context) ([]byte, error) {
-	if b.buf == nil {
-		err := b.load(ctx)
-		if err != nil {
-			return nil, err
-		}
-	}
-	return b.buf[:], nil
+	return b.bytes(ctx)
 }
 
 func (b *ByteArray) ToString(ctx context.Context) (string, error) {
-	if b.buf == nil {
-		err := b.load(ctx)
-		if err != nil {
-			return "", err
-		}
+	buf, err := b.bytes(ctx)
+	if err != nil {
+		return "", err
+	}
+	toShow := bytePeekLength
+	if len(buf) < toShow {
+		toShow = len(buf)
+	}
+	return string(buf[:toShow]), nil
+}
+
+type JSONArray struct {
+	ImmutableTree
+}
+
+func NewJSONArray(addr hash.Hash, ns NodeStore) *JSONArray {
+	return &JSONArray{ImmutableTree{Addr: addr, ns: ns}}
+}
+
+func (b *JSONArray) ToJSONDocument(ctx context.Context) (sql.JSONDocument, error) {
+	buf, err := b.bytes(ctx)
+	if err != nil {
+		return sql.JSONDocument{}, err
+	}
+	var doc sql.JSONDocument
+	err = json.Unmarshal(buf, &doc.Val)
+	if err != nil {
+		return sql.JSONDocument{}, err
 	}
-	toShow := 128
-	if len(b.buf) < toShow {
-		toShow = len(b.buf)
+	return doc, err
+}
+
+func (b *JSONArray) ToString(ctx context.Context) (string, error) {
+	buf, err := b.bytes(ctx)
+	if err != nil {
+		return "", err
 	}
-	return string(b.buf[:toShow]), nil
+	toShow := bytePeekLength
+	if len(buf) < toShow {
+		toShow = len(buf)
+	}
+	return string(buf[:toShow]), nil
+}
+
+type StringArray struct {
```

Comment:
Can we call this TextStorage or something? The type isn't an array, it's a string type that stores its data as an array